File: android/sample-ui/src/test/kotlin/com/usesmileid/sampleapps/ui/UseSmileIDSampleLaunchArgsSpecTest.kt
package com.usesmileid.sampleapps.ui

import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleFlowRoute
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleProduct
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleScenario
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleHoldCamera
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleLaunchArgs
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class UseSmileIDSampleLaunchArgsSpecTest {

    private val specArgs: List<Pair<String, String?>> by lazy {
        NAME_AND_DEFAULT.findAll(spec("launch-args.json"))
            .map { it.groupValues[1] to it.groupValues[2].takeIf { raw -> raw != "null" }?.trim('"') }
            .toList()
    }

    @Test
    fun the_canonical_names_match_the_spec() {
        assertTrue("extracted no launch arguments", specArgs.isNotEmpty())
        assertEquals(specArgs.map { it.first }, UseSmileIDSampleLaunchArgs.names)
    }

    @Test
    fun the_defaults_match_the_spec() {
        val defaults = UseSmileIDSampleLaunchArgs()
        assertEquals(
            specArgs.toMap(),
            mapOf(
                UseSmileIDSampleLaunchArgs.SCENARIO to defaults.scenario.id,
                UseSmileIDSampleLaunchArgs.THEME to defaults.theme.id,
                UseSmileIDSampleLaunchArgs.ROUTE to defaults.route.id,
                UseSmileIDSampleLaunchArgs.AUTOSTART to defaults.autostart?.id,
                UseSmileIDSampleLaunchArgs.SEED_JOBS to defaults.seedJobs.toString(),
                UseSmileIDSampleLaunchArgs.PROBES to defaults.probes.toString(),
                UseSmileIDSampleLaunchArgs.APP_LOCALE to defaults.appLocale,
                UseSmileIDSampleLaunchArgs.HOLD_CAMERA to defaults.holdCamera?.toString(),
            ),
        )
    }

    @Test
    fun an_empty_launch_is_the_declared_defaults() {
        assertEquals(UseSmileIDSampleLaunchArgs(), UseSmileIDSampleLaunchArgs.from(emptyMap()))
    }

    @Test
    fun every_argument_is_read_from_its_canonical_name() {
        val args = UseSmileIDSampleLaunchArgs.from(
            mapOf(
                UseSmileIDSampleLaunchArgs.SCENARIO to "expiredToken",
                UseSmileIDSampleLaunchArgs.THEME to "clashingHost",
                UseSmileIDSampleLaunchArgs.ROUTE to "shell",
                UseSmileIDSampleLaunchArgs.AUTOSTART to "biometricKyc",
                UseSmileIDSampleLaunchArgs.PROBES to true,
                UseSmileIDSampleLaunchArgs.APP_LOCALE to "fr-FR",
                UseSmileIDSampleLaunchArgs.HOLD_CAMERA to "keep",
            ),
        )
        assertTrue(args.probes)
        assertEquals(UseSmileIDSampleScenario.ExpiredToken, args.scenario)
        assertEquals(UseSmileIDSampleFlowRoute.Shell, args.route)
        assertEquals(UseSmileIDSampleProduct.BiometricKyc, args.autostart)
        assertEquals("fr-FR", args.appLocale)
        assertEquals(UseSmileIDSampleHoldCamera.Keep, args.holdCamera)
    }

    @Test
    fun the_retired_sandbox_argument_is_neither_declared_nor_read() {
        assertFalse("sandbox" in UseSmileIDSampleLaunchArgs.names)
        assertEquals(
            UseSmileIDSampleLaunchArgs(),
            UseSmileIDSampleLaunchArgs.from(mapOf("sandbox" to false)),
        )
    }

    @Test
    fun the_boolean_arguments_read_either_extra_type() {
        assertTrue(UseSmileIDSampleLaunchArgs.from(mapOf("seedJobs" to true)).seedJobs)
        assertTrue(UseSmileIDSampleLaunchArgs.from(mapOf("seedJobs" to "TRUE")).seedJobs)
        // A deep link's query parameter arrives as a String, which is the only form that path has.
        assertTrue(UseSmileIDSampleLaunchArgs.from(mapOf("probes" to "true")).probes)
        assertTrue(UseSmileIDSampleLaunchArgs.from(mapOf("probes" to true)).probes)
        assertFalse(UseSmileIDSampleLaunchArgs.from(mapOf("probes" to "yes")).probes)
    }

    @Test
    fun hold_camera_accepts_milliseconds_as_well_as_keep() {
        assertEquals(
            UseSmileIDSampleHoldCamera.Millis(1500),
            UseSmileIDSampleLaunchArgs.from(mapOf("holdCamera" to "1500")).holdCamera,
        )
        assertNull(UseSmileIDSampleLaunchArgs.from(mapOf("holdCamera" to "soon")).holdCamera)
    }

    @Test
    fun an_unrecognised_value_falls_back_to_its_default() {
        val args = UseSmileIDSampleLaunchArgs.from(mapOf("scenario" to "typo", "route" to "sheet"))
        assertEquals(UseSmileIDSampleScenario.Normal, args.scenario)
        assertEquals(UseSmileIDSampleFlowRoute.Fullscreen, args.route)
    }

    private companion object {
        /** Non-greedy to the first `default` after each name, so an arg's own `values` list is skipped. */
        val NAME_AND_DEFAULT =
            Regex("\"name\"\\s*:\\s*\"(\\w+)\"[\\s\\S]*?\"default\"\\s*:\\s*(\"[^\"]*\"|null|true|false)")
    }
}

File: flutter/sample_ui/test/spec/use_smileid_sample_launch_args_spec_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sample_ui/sample_ui.dart';

import 'spec_file.dart';

void main() {
  late List<Map<String, Object?>> args;

  setUpAll(() => args = objects(spec('launch-args.json')['args']));

  test('the canonical names match the spec', () {
    final List<String> expected = args
        .map((Map<String, Object?> it) => it['name']! as String)
        .toList();
    expect(expected, isNotEmpty, reason: 'extracted no launch arguments');
    expect(UseSmileIDSampleLaunchArgs.names, expected);
  });

  test('the defaults match the spec', () {
    const UseSmileIDSampleLaunchArgs defaults = UseSmileIDSampleLaunchArgs();
    final Map<String, Object?> ours = <String, Object?>{
      UseSmileIDSampleLaunchArgs.scenarioArg: defaults.scenario.id,
      UseSmileIDSampleLaunchArgs.themeArg: defaults.theme.id,
      UseSmileIDSampleLaunchArgs.routeArg: defaults.route.id,
      UseSmileIDSampleLaunchArgs.autostartArg: defaults.autostart?.id,
      UseSmileIDSampleLaunchArgs.seedJobsArg: defaults.seedJobs,
      UseSmileIDSampleLaunchArgs.seedProfilesArg: defaults.seedProfiles,
      UseSmileIDSampleLaunchArgs.probesArg: defaults.probes,
      UseSmileIDSampleLaunchArgs.appLocaleArg: defaults.appLocale,
      UseSmileIDSampleLaunchArgs.holdCameraArg: defaults.holdCamera,
      UseSmileIDSampleLaunchArgs.noticeWindowArg: defaults.noticeWindow,
    };
    final Map<String, Object?> theirs = <String, Object?>{
      for (final Map<String, Object?> arg in args)
        arg['name']! as String: arg['default'],
    };
    expect(ours, theirs);
  });

  test('an empty launch is the declared defaults', () {
    expect(
      UseSmileIDSampleLaunchArgs.from(const <String, Object?>{}),
      const UseSmileIDSampleLaunchArgs(),
    );
  });

  test('a launch with no arguments seeds nothing', () {
    const UseSmileIDSampleLaunchArgs plain = UseSmileIDSampleLaunchArgs();
    expect(plain.seedJobs, isFalse);
    expect(plain.seedProfiles, isFalse);
    expect(plain.probes, isFalse);
    expect(plain.autostart, isNull);
  });

  test('every argument is read from its canonical name', () {
    final UseSmileIDSampleLaunchArgs parsed =
        UseSmileIDSampleLaunchArgs.from(const <String, Object?>{
          UseSmileIDSampleLaunchArgs.scenarioArg: 'expiredToken',
          UseSmileIDSampleLaunchArgs.themeArg: 'clashingHost',
          UseSmileIDSampleLaunchArgs.routeArg: 'shell',
          UseSmileIDSampleLaunchArgs.autostartArg: 'biometricKyc',
          UseSmileIDSampleLaunchArgs.seedJobsArg: true,
          UseSmileIDSampleLaunchArgs.seedProfilesArg: true,
          UseSmileIDSampleLaunchArgs.probesArg: true,
          UseSmileIDSampleLaunchArgs.appLocaleArg: 'fr-FR',
          UseSmileIDSampleLaunchArgs.holdCameraArg: 'keep',
          UseSmileIDSampleLaunchArgs.noticeWindowArg: '60',
        });
    expect(parsed.scenario, UseSmileIDSampleScenario.expiredToken);
    expect(parsed.theme, UseSmileIDSampleThemeScenario.clashingHost);
    expect(parsed.route, UseSmileIDSampleFlowRoute.shell);
    expect(parsed.autostart, UseSmileIDSampleProduct.biometricKyc);
    expect(parsed.seedJobs, isTrue);
    expect(parsed.seedProfiles, isTrue);
    expect(parsed.probes, isTrue);
    expect(parsed.appLocale, 'fr-FR');
    expect(parsed.holdCamera, const UseSmileIDSampleHoldCameraKeep());
    expect(parsed.noticeWindow, 60);
  });

  test(
    'a cold-start link carries the same nine names through its query string',
    () {
      final UseSmileIDSampleLaunchArgs
      parsed = UseSmileIDSampleLaunchArgs.fromUri(
        Uri.parse(
          'usesmileid-sample-flutter://run?scenario=badRefresh&route=shell&seedJobs=true',
        ),
      );
      expect(parsed.scenario, UseSmileIDSampleScenario.badRefresh);
      expect(parsed.route, UseSmileIDSampleFlowRoute.shell);
      expect(parsed.seedJobs, isTrue);
      expect(parsed.seedProfiles, isFalse);
    },
  );

  test('the boolean arguments read either form a launch can deliver', () {
    bool seeds(Object? value) => UseSmileIDSampleLaunchArgs.from(
      <String, Object?>{'seedJobs': value},
    ).seedJobs;
    expect(seeds(true), isTrue);
    expect(seeds('TRUE'), isTrue);
    expect(seeds('false'), isFalse);
    expect(seeds('yes'), isFalse);
  });

  test('notice window takes positive seconds only', () {
    int? window(String value) => UseSmileIDSampleLaunchArgs.from(
      <String, Object?>{'noticeWindow': value},
    ).noticeWindow;
    expect(window('60'), 60);
    expect(window('soon'), isNull);
    expect(window('0'), isNull);
    expect(window('-5'), isNull);
  });

  test('hold camera accepts milliseconds as well as keep', () {
    expect(
      UseSmileIDSampleLaunchArgs.from(const <String, Object?>{
        'holdCamera': '1500',
      }).holdCamera,
      const UseSmileIDSampleHoldCameraMillis(1500),
    );
    expect(
      UseSmileIDSampleLaunchArgs.from(const <String, Object?>{
        'holdCamera': 'soon',
      }).holdCamera,
      isNull,
    );
  });

  test('an unrecognised value falls back to its default', () {
    final UseSmileIDSampleLaunchArgs parsed = UseSmileIDSampleLaunchArgs.from(
      const <String, Object?>{'scenario': 'typo', 'route': 'sheet'},
    );
    expect(parsed.scenario, UseSmileIDSampleScenario.normal);
    expect(parsed.route, UseSmileIDSampleFlowRoute.fullscreen);
  });

  test('the retired sandbox argument is neither declared nor read', () {
    expect(UseSmileIDSampleLaunchArgs.names, isNot(contains('sandbox')));
    expect(
      UseSmileIDSampleLaunchArgs.from(const <String, Object?>{
        'sandbox': false,
      }),
      const UseSmileIDSampleLaunchArgs(),
    );
  });
}
